File: iosApp/iosApp/AppStore.swift
//
//  AppStore.swift
//  KMP Redux
//
//  Created by Clément Beffa on 20.02.22.
//

import Foundation
import shared
import Combine
import KMPNativeCoroutinesCombine

class AppStore: ObservableObject {
    @Published private(set) var state: AppState
    private let store = shared.AppStore()
    private var cancellable: AnyCancellable?
    
    
    init() {
        state = store.stateNativeValue
        cancellable = createPublisher(for: store.stateNative)
            .receive(on: DispatchQueue.main)
            .sink(receiveCompletion: { _ in
                
            }, receiveValue: { state in
                self.state = state
            })
    }
    
    func dispatch(_ action: AppActionKs) {
        store.dispatch(action.sealed)
    }
        
    func derived<DerivedState: Equatable, ExtractedAction>(
        deriveState: @escaping (AppState) -> DerivedState,
        embedAction: @escaping (ExtractedAction) -> AppActionKs
    ) -> DerivedStore<DerivedState, ExtractedAction> {
        let store = DerivedStore<DerivedState, ExtractedAction>(
            initialState: deriveState(state),
            dispatch: { action  in
                self.dispatch(embedAction(action))
            }
        )

        $state
            .map(deriveState)
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .assign(to: &store.$state)

        return store
    }
}

class DerivedStore<State, Action>: ObservableObject {
    @Published fileprivate(set) var state: State
    let dispatch: (Action) -> Void
    
    init(initialState: State, dispatch: @escaping (Action) -> Void) {
        self.state = initialState
        self.dispatch = dispatch
    }
}

File: shared/src/commonMain/kotlin/kmp/redux/redux/AppStore.kt
package kmp.redux.redux

import kotlinx.coroutines.MainScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

class AppStore {
    companion object {
        val shared = AppStore()
    }
    private val stateFlow = MutableStateFlow(AppState())
    lateinit var dispatch: (AppAction) -> Unit

    init {
        dispatch = {
            stateFlow.value = appReducer(stateFlow.value, it) {
            this.dispatch(it)
        }
    }
    }

    val state: StateFlow<AppState>
        get() = stateFlow

    fun <DerivedState, ExtractedAction> derived(deriveState: (AppState) -> DerivedState, embedAction: (ExtractedAction) -> AppAction): DerivedStore<DerivedState, ExtractedAction> {
        val store = DerivedStore<DerivedState, ExtractedAction>(deriveState(state.value)) {
            dispatch(embedAction(it))
        }

        MainScope().launch {
            state.collect {
                store.stateFlow.value = deriveState(it)
            }
        }

        return store
    }
}

class DerivedStore<State, Action>(initialState: State, val dispatch: (Action) -> Unit) {
    internal val stateFlow: MutableStateFlow<State> = MutableStateFlow(initialState)

    val state: StateFlow<State>
        get() = stateFlow
}